docs/developers/how-to-build-plugins.md

```diff
@@ -0,0 +1,262 @@
+To help developers write Typescript models to integrate easily into IF, we provide the `ModelPluginInterface` interface. Here's an overview of the stages you need to follow to integrate your model:
+
+- create a Typescript file that implements the `ModelPluginInterface`
+- install the model
+- initialize and invoke the model in your manifest file
+
+
+## The model interface
+
+The `ModelPluginInterface` is structured as follows:
+
+```ts
+export interface ModelPluginInterface {
+  /**
+   * Configures instance with given params.
+   */
+  configure(params: object | undefined): Promise<ModelPluginInterface>;
+
+  /**
+   * Calculates `output` based on given model's `input`.
+   */
+  execute(inputs: ModelParams[]): Promise<ModelParams[]>;
+}
+```
+
+## Methods
+
+### configure
+
+`Configure()` applies any model configuration required by the model, such as setting values of global constants during model initialization.
+
+#### Params
+
+| Param    | Type                  | Purpose                                          |
+| -------- | --------------------- | ------------------------------------------------ |
+| `params` | `object or undefined` | optional model config data passed as an `object` |
+
+#### Returns
+
+| Return value | Type                            | Purpose                                                          |
+| ------------ | ------------------------------- | ---------------------------------------------------------------- |
+| `model`      | `Promise<ModelPluginInterface>` | `Promise` resolving to an instance of the `ModelPluginInterface` |
+
+
+
+### execute
+
+`execute()` is where the main calculation logic of the model is implemented.
+
+#### Params
+
+| Param    | Type            | Purpose                                                                        |
+| -------- | --------------- | ------------------------------------------------------------------------------ |
+| `inputs` | `ModelParams[]` | Array of data provided in the `inputs` field of a component in a manifest file |
+
+
+#### Returns
+
+| Return value | Type                     | Purpose                                                    |
+| ------------ | ------------------------ | ---------------------------------------------------------- |
+| `outputs`    | `Promise<ModelParams[]>` | `Promise` resolving to an array of updated `ModelParams[]` |
+
+
+## What are `ModelParams`?
+
+`ModelParams` are a fundamental data type in the Impact Framework.  The type is defined as follows:
+
+```ts
+export type ModelParams = {
+  [K in UnitKeyName]?: any;
+};
+```
+
+The `ModelParams` type therefore defines an array of key-value pairs. The keys must exist in the `UnitKeys` array in `units.ts`, so if your new model uses some novel input values or generates some metric that hasn't been used in IF before, then you will have to add them to that array.
```

Comment:
Interesting I wasn't aware of this change. Does this mean that as well as updating units.yml if you are creating a new param you also have to update this additional TS file?

Is there a reason for this constraint? 

It also breaks the design philosophy of decentralisation since for every new param it has to be added to the units.ts in the CLI repo.

Reply:
yes - we know this validation step is going to break when we update `units.yml`, so you can consider this particular piece of logic deprecated but not yet removed.